packages/databricks-vscode/src/bundle/BundleProjectManager.ts

```diff
@@ -0,0 +1,143 @@
+import {
+    QuickPickItem,
+    QuickPickItemKind,
+    Disposable,
+    Uri,
+    window,
+    commands,
+} from "vscode";
+import {ConnectionManager} from "../configuration/ConnectionManager";
+import {ConfigModel} from "../configuration/models/ConfigModel";
+import {BundleFileSet} from "./BundleFileSet";
+import {logging} from "@databricks/databricks-sdk";
+import {Loggers} from "../logger";
+import {CachedValue} from "../locking/CachedValue";
+import {CustomWhenContext} from "../vscode-objs/CustomWhenContext";
+
+export class BundleProjectManager {
+    private logger = logging.NamedLogger.getOrCreate(Loggers.Extension);
+    private disposables: Disposable[] = [];
+
+    private _isBundleProject = new CachedValue<boolean>(async () => {
+        const rootBundleFile = await this.bundleFileSet.getRootFile();
+        return rootBundleFile !== undefined;
+    });
+
+    public onDidChangeStatus = this._isBundleProject.onDidChange;
+
+    private _isLegacyProject = new CachedValue<boolean>(async () => {
+        // TODO
+        return false;
+    });
+
+    private _subProjects = new CachedValue<{absolute: Uri; relative: Uri}[]>(
+        async () => {
+            const subProjects = await this.bundleFileSet.getSubProjects();
+            this.customWhenContext.setSubProjectsAvailable(
+                subProjects?.length > 0
+            );
+            return subProjects;
+        }
+    );
+
+    constructor(
+        private customWhenContext: CustomWhenContext,
+        private connectionManager: ConnectionManager,
+        private configModel: ConfigModel,
+        private bundleFileSet: BundleFileSet,
+        private workspaceUri: Uri
+    ) {}
+
+    public async isBundleProject(): Promise<boolean> {
+        return await this._isBundleProject.value;
+    }
+
+    public async configureWorkspace(): Promise<void> {
```

Comment:
We should maybe try to reconfigure when _isBundleProject changes?

Reply:
Makes sense, for that we need to move a lot of initialisation logic into the project manager so we can properly dispose it during "reconfiguration" process. Let me do it in a follow up after we have the init wizard in a terminal

Reply:
Improved a bit the initialization flow in the new revision